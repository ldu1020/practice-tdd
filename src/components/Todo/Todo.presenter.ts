class TodoPresenter {
  list: Item[] = [];
  constructor(initialList: Item[]) {
    this.list = [...initialList];
  }

  addItem(item: Item, updateFn?: UpdateFn) {
    this.list = [...this.list, item];
    if (updateFn) updateFn(this.list);
  }

  deleteItem(id: number, updateFn?: UpdateFn) {
    const updated = this.list.filter((li) => li.id !== id);
    this.list = updated;
    if (updateFn) updateFn(this.list);
  }

  checkItem(id: number, updateFn?: UpdateFn) {
    const updated = this.list.map((li) => {
      if (li.id === id) li.checked = !li.checked;
      return li;
    });
    this.list = updated;
    if (updateFn) updateFn(this.list);
  }

  getList() {
    return this.list;
  }

  get isEmpty() {
    return this.list.length === 0;
  }
}

export default TodoPresenter;

export type Item = {
  id: number;
  title: string;
  checked: boolean;
};

export type UpdateFn = (items: Item[]) => void;
