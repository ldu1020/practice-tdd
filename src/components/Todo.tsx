import React from "react";
import { Flex, List, ListIcon, ListItem } from "@chakra-ui/layout";
import TodoPresenter, { Item } from "./Todo/Todo.presenter";
import { CheckIcon, CloseIcon } from "@chakra-ui/icons";
import TodoAddForm from "./Todo/TodoAddForm/TodoAddForm";

interface TodoProps {
  todoPresenter: TodoPresenter;
}

const Todo = ({ todoPresenter }: TodoProps) => {
  const [list, setList] = React.useState(todoPresenter.getList());
  const handleAddTodo = (item: Item) => {
    todoPresenter.addItem(item, setList);
  };

  const handleDeleteTodo = (id: number) => {
    todoPresenter.deleteItem(id, setList);
  };

  const handleCheckTodo = (id: number) => {
    todoPresenter.checkItem(id, setList);
  };

  return (
    <List w="300px" border="1px">
      <TodoAddForm onAdd={handleAddTodo} />
      {list.map((item) => {
        return (
          <ListItem
            key={item.id}
            display="flex"
            justifyContent="space-between"
            alignItems="center"
            borderBottom="1px"
            p="10px"
            cursor="pointer"
            onClick={() => {
              handleCheckTodo(item.id);
            }}
          >
            {item.title}
            <Flex>
              <ListIcon as={CheckIcon} color={item.checked ? "red" : "gray"} />
              <ListIcon
                as={CloseIcon}
                onClick={() => {
                  handleDeleteTodo(item.id);
                }}
              />
            </Flex>
          </ListItem>
        );
      })}
    </List>
  );
};

export default Todo;
