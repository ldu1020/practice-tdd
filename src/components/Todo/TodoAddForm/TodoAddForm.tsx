import React from "react";
import { Item } from "../Todo.presenter";
import { Button, Input } from "@chakra-ui/react";

interface TodoAddFormProps {
  onAdd: (item: Item) => void;
}

const TodoAddForm = ({ onAdd }: TodoAddFormProps) => {
  const inputRef = React.useRef<HTMLInputElement | null>(null);
  const formRef = React.useRef<HTMLFormElement | null>(null);

  const onSubmit: React.MouseEventHandler<HTMLButtonElement> = (e) => {
    e.preventDefault();
    const input = inputRef.current;
    const form = formRef.current;
    const title = input?.value;
    if (title) {
      onAdd({ id: Date.now(), title, checked: false });
      form?.reset();
    }
  };
  return (
    <form ref={formRef}>
      <Input ref={inputRef} placeholder="item" />
      <Button onClick={onSubmit}>Add</Button>
    </form>
  );
};

export default TodoAddForm;
