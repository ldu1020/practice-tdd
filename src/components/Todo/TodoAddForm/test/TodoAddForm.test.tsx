import { render, screen } from "@testing-library/react";
import renderer from "react-test-renderer";
import { Item } from "../../Todo.presenter";
import TodoAddForm from "../TodoAddForm";
import userEvent from "@testing-library/user-event";

describe("TodoAddForm", () => {
  it("renders (snapshot)", () => {
    const components = renderer.create(<TodoAddForm onAdd={jest.fn()} />);
    expect(components.toJSON()).toMatchSnapshot();
  });

  describe("Form Submit", () => {
    let onAdd: (item: Item) => void;
    let button: HTMLButtonElement;
    let input: HTMLInputElement;

    beforeEach(() => {
      onAdd = jest.fn();
      // eslint-disable-next-line testing-library/no-render-in-setup
      render(<TodoAddForm onAdd={onAdd} />);
      button = screen.getByText("Add");
      input = screen.getByPlaceholderText("item");
    });

    it('calls "onAdd" when buttons is clicked and valid value', () => {
      userEvent.type(input, "new item");
      userEvent.click(button);
      expect(onAdd).toHaveBeenCalled();
    });

    it('does not call "onAdd" when buttons is clicked and empty value', () => {
      userEvent.type(input, "");
      userEvent.click(button);
      expect(onAdd).not.toHaveBeenCalled();
    });
  });
});
