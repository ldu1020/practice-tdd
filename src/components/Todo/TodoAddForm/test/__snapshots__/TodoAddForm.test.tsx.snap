// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoAddForm renders (snapshot) 1`] = `
<form>
  <input
    className="chakra-input css-0"
    onBlur={[Function]}
    onFocus={[Function]}
    placeholder="item"
  />
  <button
    className="chakra-button css-5c2176"
    onClick={[Function]}
    type="button"
  >
    Add
  </button>
</form>
`;
